In [16]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

In [17]:
# Load the CSV file containing responses
df = pd.read_csv('Dataset/encoded_response.csv')

In [18]:
# # Target variable: Learning Objects Preference
target = df[[
    'Learning Objects [Slide presentation]',
    'Learning Objects [Book]',
    'Learning Objects [Lecture Note]',
    'Learning Objects [Educational game]',
    'Learning Objects [Video]',
    'Learning Objects [Audio-recorded lecture]',
    'Learning Objects [Animated instruction]',
    'Learning Objects [Real object model]',
    'Learning Objects [Mind Map]',
    'Learning Objects [Multimedia content]',
    'Learning Objects [Interactive Tool]',
    'Learning Objects [Technology-supported learning include computer-based training systems]',
    'Learning Objects [Intelligent computer-aided instruction systems]'
]]

In [19]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df.drop(target.columns, axis=1), target, test_size=0.25, random_state=42)

In [20]:
# # Create a RandomForestClassifier for each learning object
# classifiers = {}
# for col in target.columns:
#     classifier = RandomForestClassifier(n_estimators=100, random_state=42)
#     classifier.fit(X_train, y_train[col])
#     classifiers[col] = classifier

In [21]:
from sklearn.model_selection import GridSearchCV
import joblib

In [22]:
# # Create an empty dictionary to hold the classifiers
# best_estimators = {}

# # Loop through each learning object
# for col in target.columns:
#     # Define the parameter grid for hyperparameter tuning
#     param_grid = {
#         'n_estimators': [100, 150, 200],  # Vary the number of trees
#         'max_depth': [None, 10, 20, 30],  # Vary the maximum depth of trees
#         # Add other hyperparameters to tune
#     }
    
#     # Instantiate GridSearchCV for RandomForestClassifier
#     grid_search = GridSearchCV(RandomForestClassifier(random_state=42), param_grid, cv=5, scoring='accuracy')
    
#     # Fit the grid search to your data for the current learning object
#     grid_search.fit(X_train, y_train[col])
    
#     # Get the best parameters and best estimator for the current learning object
#     best_params = grid_search.best_params_
#     best_estimator = grid_search.best_estimator_
    
#     # Store the best estimator in the classifiers dictionary
#     best_estimators[col] = best_estimator
    
# joblib.dump(best_estimators, "Model/rf_model_2.joblib")

In [23]:
# Create an empty dictionary to hold the classifiers
best_estimators = {}

# Loop through each learning object
for col in target.columns:
    # Define the parameter grid for hyperparameter tuning
    param_grid = {
        'n_estimators': [100, 150, 200, 250],  # Expanded number of trees
        'max_depth': [None, 10, 20, 30, 50],  # Expanded maximum depth of trees
        'min_samples_split': [2, 5, 10],  # Minimum samples for node splitting
        'min_samples_leaf': [1, 2, 4, 8],  # Minimum samples required at leaf nodes
        'max_features': ['auto', 'sqrt', 'log2']  # Maximum number of features considered for splitting
        # Add other hyperparameters to tune
    }
    
    # Instantiate GridSearchCV for RandomForestClassifier
    grid_search = GridSearchCV(RandomForestClassifier(random_state=42), param_grid, cv=5, scoring='accuracy')
    
    # Fit the grid search to your data for the current learning object
    grid_search.fit(X_train, y_train[col])
    
    # Get the best parameters and best estimator for the current learning object
    best_params = grid_search.best_params_
    best_estimator = grid_search.best_estimator_
    
    # Store the best estimator in the classifiers dictionary
    best_estimators[col] = best_estimator
    
    # Evaluate the model on the validation set
    y_pred = best_estimator.predict(X_test)
    accuracy = accuracy_score(y_test[col], y_pred)
    print(f"Model Accuracy for {col}: {accuracy}")

    
# Save the best estimators using joblib
joblib.dump(best_estimators, "Model/rf_model_3.joblib")

C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Slide presentation]: 0.5868725868725869


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Book]: 0.722007722007722


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Lecture Note]: 0.7027027027027027


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Educational game]: 0.6486486486486487


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Video]: 0.6061776061776062


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Audio-recorded lecture]: 0.7065637065637066


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Animated instruction]: 0.5907335907335908


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Real object model]: 0.6486486486486487


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Mind Map]: 0.6718146718146718


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Multimedia content]: 0.5907335907335908


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Interactive Tool]: 0.6254826254826255


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Technology-supported learning include computer-based training systems]: 0.5675675675675675


C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py:425: FitFailedWarning: 
1200 fits failed out of a total of 3600.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1200 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\model_selection\_validation.py", line 729, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 1145, in wrapper
    estimator._validate_params()
  File "C:\Users\Janice\anaconda3\envs\dspenv\lib\site-packages\sklearn\base.py", line 638, in _validate_params
    validate_parameter_constr

Model Accuracy for Learning Objects [Intelligent computer-aided instruction systems]: 0.637065637065637


['Model/rf_model_3.joblib']

In [24]:
print('best_params:', best_params)

best_params: {'max_depth': 10, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 5, 'n_estimators': 150}


In [25]:
# Make predictions on the testing set for each learning object
y_pred = pd.DataFrame({col: classifier.predict(X_test) for col, classifier in best_estimators.items()})

sum_acc = 0
mean_acc = 0

# Evaluate the model for each learning object
accuracy = {col: accuracy_score(y_test[col], y_pred[col]) for col in target.columns}
print("Model Accuracy for Learning Objects:")
for col, acc in accuracy.items():
    print(f"{col}: {acc}")
    sum_acc += acc
    
mean_acc = sum_acc/13
print("Mean accuracy: ", mean_acc)

Model Accuracy for Learning Objects:
Learning Objects [Slide presentation]: 0.5868725868725869
Learning Objects [Book]: 0.722007722007722
Learning Objects [Lecture Note]: 0.7027027027027027
Learning Objects [Educational game]: 0.6486486486486487
Learning Objects [Video]: 0.6061776061776062
Learning Objects [Audio-recorded lecture]: 0.7065637065637066
Learning Objects [Animated instruction]: 0.5907335907335908
Learning Objects [Real object model]: 0.6486486486486487
Learning Objects [Mind Map]: 0.6718146718146718
Learning Objects [Multimedia content]: 0.5907335907335908
Learning Objects [Interactive Tool]: 0.6254826254826255
Learning Objects [Technology-supported learning include computer-based training systems]: 0.5675675675675675
Learning Objects [Intelligent computer-aided instruction systems]: 0.637065637065637
Mean accuracy:  0.6388476388476388


test_size=0.1 --    
test_size=0.2 -- 0.7729 -- 0.6386  0.63647  
model 3= 0.6388

### Make predictions on a new data

In [5]:
rf_model = joblib.load("Model/rf_model.joblib")

In [6]:
data = pd.read_csv("Dataset/data.csv")

In [7]:
predictions = pd.DataFrame({col: classifier.predict(data) for col, classifier in rf_model.items()})
predictions.head()

Learning Objects [Slide presentation]  Learning Objects [Book]  \
0                                      1                        0   
1                                      1                        1   
2                                      1                        0   
3                                      0                        0   
4                                      1                        0   

   Learning Objects [Lecture Note]  Learning Objects [Educational game]  \
0                                1                                    0   
1                                1                                    1   
2                                1                                    0   
3                                1                                    0   
4                                1                                    0   

   Learning Objects [Video]  Learning Objects [Audio-recorded lecture]  \
0                         0                                          1   
1                         1                                          0   
2                         1                                          1   
3                         0                                          1   
4                         0                                          0   

   Learning Objects [Animated instruction]  \
0                                        1   
1                                        0   
2                                        1   
3                                        1   
4                                        0   

   Learning Objects [Real object model]  Learning Objects [Mind Map]  \
0                                     1                            1   
1                                     0                            1   
2                                     1                            1   
3                                     1                            0   
4                                     1                            0   

   Learning Objects [Multimedia content]  Learning Objects [Interactive Tool]  \
0                                      0                                    0   
1                                      1                                    1   
2                                      1                                    0   
3                                      0                                    0   
4                                      0                                    1   

   Learning Objects [Technology-supported learning include computer-based training systems]  \
0                                                  0                                          
1                                                  1                                          
2                                                  0                                          
3                                                  0                                          
4                                                  0                                          

   Learning Objects [Intelligent computer-aided instruction systems]  
0                                                  0                  
1                                                  1                  
2                                                  0                  
3                                                  1                  
4                                                  0